toolz/itertoolz.py

```diff
@@ -896,3 +897,70 @@ def peek(seq):
     iterator = iter(seq)
     item = next(iterator)
     return item, itertools.chain([item], iterator)
+
+
+class PeekableIterator(object):
+    """An iterator that allows us to peek at elements without consuming them
+    from the stream.
+
+    Parameters
+    ----------
+    stream : iterator
+        The iterator to wrap.
+    """
+    def __init__(self, stream):
+        self._stream = iter(stream)
+        self._peeked = Queue()
```

Comment:
I would recommend using `collections.deque`.  `Queue` is intended for multithreading.